[FEATURE] [needs-docs] Add  the grid frame OFFSET  In the "Print Layout" map grid

konst555 opened the pull request:
## Description

Proposed changes.
In the "Print Layout" section when building a map grid, it is necessary to add the possibility of a frame shift relative to the map.
It will be very easy to form a different framework for drawing up maps.

---------
What changed :
1. UI - add label "offset" and QgsDoubleSpinBox  
2. "offset" - from 0 to 99 ( default -0)
3. Add corners for "zebra" and frame boarder.
4. If you use the default offset to "0", there will be no changes from the current version. Only for "zebra" corners will be drawn more correctly.
----------

Why it is very important to make this small change.
Now QGIS does not have the means to design a standard scale frame of nautical charts. It is not possible to make standard designs for topographic maps. 
Some applications, such as ArcGIS or my plugin for QGIS - the geogrid, create static vector layers for the map sections. When making the proposed add-on, we can very easily create a dynamic design that will meet the requirements of the standards.

![default](https://user-images.githubusercontent.com/17308458/44768713-d9936a00-ab6a-11e8-8940-16ded55f57e1.png)

![default](https://user-images.githubusercontent.com/17308458/44769138-4bb87e80-ab6c-11e8-8bf6-ed96f77eaf4e.png)


## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

DelazJ commented:
About the GUI i think it'd be better to have a proper `Frame offset` line instead of two independant properties on the same line as shown in the screenshot.

konst555 commented:
About the GUI. 
I tried to arrange "offset" on a separate line, but it was not comfortable and not logical.
+ I think that the term "size" on this line is not correct. More correct is "Frame width". In this case, the string will be logical and correct "Frame width / offset  [  ]  [  ]".
But maybe I do not know enough English well, so I left it as it is.

konst555 commented:
OK. 
I changed the GUI to the style of the old layout.
I placed the parameter on a separate line and corrected my errors.


![default](https://user-images.githubusercontent.com/17308458/44836391-7c1d1d00-ac3f-11e8-9309-7c3d08f89c19.png)

nyalldawson commented on src/core/layout/qgslayoutitemmapgrid.cpp:
```diff
@@ -963,19 +976,50 @@ void QgsLayoutItemMapGrid::drawGridFrameLineBorder( QPainter *p, QgsLayoutItemMa
   p->setBrush( Qt::NoBrush );
   p->setPen( framePen );
 
+  double X1;
```
These declarations should be moved into the scopes where the variables are used

nyalldawson commented on src/core/layout/qgslayoutitemmapgrid.cpp:
```diff
@@ -1684,7 +1728,7 @@ int QgsLayoutItemMapGrid::xGridLinesCrsTransform( const QgsRectangle &bbox, cons
       {
         Q_UNUSED( cse );
-        QgsDebugMsg( QStringLiteral( "Caught CRS exception %1" ).arg( cse.what() ) );
+        QgsDebugMsg( QString( "Caught CRS exception %1" ).arg( cse.what() ) );
```
Leave as "QStringLiteral"

nyalldawson commented:
Looking good! There's a few changes we need to do to get this merge-able.

1. I think either "padding" or "margin" is a more appropriate name here than "offset" ("offset" conveys to me a constant value applied to the whole frame, i.e. a positive offset would move the left side of the frame into the map itself). Can you update the UI for this rename and the API function names too?

2. You need to run the `scripts/prepare-commit.sh` script over the commit, in order to apply the auto-formatting and Python binding scripts. (This is usually done by setting this script as a pre-commit hook.)

3. There's some differences in the reference test images due to these changes. I'll help look into these when the other two points are addressed.

nyalldawson commented:
Oh and lastly, we'll need some new tests covering this new feature. I'll also assist with those when the other changes are done.

konst555 commented:
I have some questions about your comments.

--------------------------------------------
1) These local auxiliary variables X1 Y1 are in place before use. Where can they be moved?  Inside "switch" ?  

>In src/core/layout/qgslayoutitemmapgrid.cpp:
+  double X1;
> These declarations should be moved into the scopes where the variables are used

...
  //end point for drawLine()
  double X1;
  double Y1;

  switch ( border )
  {
    case QgsLayoutItemMapGrid::Left:
      p->drawLine( QLineF( 0-mGridFrameOffset, 0-mGridFrameOffset, 0-mGridFrameOffset, mMap->rect().height()+mGridFrameOffset ) );
      //corner left-top
      if ( mGridFrameOffset != 0 )
      {
          X1 = 0-mGridFrameOffset+mGridFramePenThickness / 2.0;
          Y1 = 0-mGridFrameOffset+mGridFramePenThickness / 2.0;
          p->drawLine( QLineF( 0, 0, X1, Y1 ) );
      }
...

--------------------------------------------
2) This name "offset" I took by analogy to the line style. For labels around the frame, "Distance to map frame" is used, but this is very long. The terms that you offer are very rarely used, so it will be very difficult for non-English speakers to translate. 
I suggest correcting the label for "Offset from the map." Do you agree with this?

>* 
>I think either "padding" or "margin" is a more appropriate name here than "offset" ("offset" conveys to me a constant value applied to the whole frame, i.e. a positive offset would move the left side of the frame into the map itself). Can you update the UI for this rename and the API function names too?
>* 

----------------------------------------------
3) I already suggested changing the label  "Frame size" to  "Frame width" .  "size" refers to the entire frame from the left to the right side of the border, but here we mean only the width by analogy with the line. When I started working with QGIS, I did not immediately understand what this "size" means.
What do you think about this? If you agree, I will correct this label.

nyalldawson commented:
> ) These local auxiliary variables X1 Y1 are in place before use. Where can they be moved?  Inside "switch" ?
In src/core/layout/qgslayoutitemmapgrid.cpp:

Yes, I'd rather see them done as:

    const double X1 = 0-mGridFrameOffset+mGridFramePenThickness / 2.0;

etc

This would avoid any chance of them being used uninitialized if this code is refactored in future.

> 2) This name "offset" I took by analogy to the line style. For labels around the frame, "Distance to map frame" is used, but this is very long. The terms that you offer are very rarely used, so it will be very difficult for non-English speakers to translate.
I suggest correcting the label for "Offset from the map." Do you agree with this?

I disagree - padding and margin are used heavily for similar functionality, e.g. in CSS. We also use margin for similar functionality within QGIS, such as the margins for a layout item label, decoration placements, annotation margins, etc. I think this should be renamed to "margin" for consistency with these other settings.


>  I already suggested changing the label  "Frame size" to  "Frame width" .  "size" refers to the entire frame from the left to the right side of the border, but here we mean only the width by analogy with the line. When I started working with QGIS, I did not immediately understand what this "size" means.
What do you think about this? If you agree, I will correct this label.

Yes, +1 to change this. Can you do this in a separate PR please? Don't forget to tag the commit message with "[needs-docs]", so that the documentation team are notified of the change.

konst555 commented:
Thanks for the help. If I did something wrong, write to me. I'll try to fix it.

I tried to make a separate branch for PR to fix "Frame size" to "Frame width", but I did not succeed.
Need to change one word in the file "qgslayoutmapgridwidgetbase.ui". Created PR is in my Fork from the main repository.

https://github.com/konst555/QGIS/pull/1  

I guess I made a mistake somewhere creating this PR, I'm sorry.


>Вторник,  4 сентября 2018, 3:08 +03:00 от Nyall Dawson <notifications@github.com>:
>
>>) These local auxiliary variables X1 Y1 are in place before use. Where can they be moved?  Inside "switch" ?
>>In src/core/layout/qgslayoutitemmapgrid.cpp:
>Yes, I'd rather see them done as:
>const double X1 = 0-mGridFrameOffset+mGridFramePenThickness / 2.0;
>
>etc
>This would avoid any chance of them being used uninitialized if this code is refactored in future.
>>*  
>>This name "offset" I took by analogy to the line style. For labels around the frame, "Distance to map frame" is used, but this is very long. The terms that you offer are very rarely used, so it will be very difficult for non-English speakers to translate.
>>I suggest correcting the label for "Offset from the map." Do you agree with this?
>I disagree - padding and margin are used heavily for similar functionality, e.g. in CSS. We also use margin for similar functionality within QGIS, such as the margins for a layout item label, decoration placements, annotation margins, etc. I think this should be renamed to "margin" for consistency with these other settings.
>>I already suggested changing the label  "Frame size" to  "Frame width" .  "size" refers to the entire frame from the left to the right side of the border, but here we mean only the width by analogy with the line. When I started working with QGIS, I did not immediately understand what this "size" means.
>>What do you think about this? If you agree, I will correct this label.
>Yes, +1 to change this. Can you do this in a separate PR please? Don't forget to tag the commit message with "[needs-docs]", so that the documentation team are notified of the change.
>—
>You are receiving this because you authored the thread.
>Reply to this email directly,  view it on GitHub , or  mute the thread .

-- 
Konstantin Puzankov
